tools/retest/main.go

```diff
@@ -23,75 +26,180 @@ func main() {
 			}
+			log.Printf("#%d retesting: %s %s", prNumber, job, state)
+			comment := github.IssueComment{
+				Body: ptr("/test " + job),
+			}
+			testComment, _, err := client.Issues.CreateComment(ctx, S, S, prNumber, &comment)
+			handleError(err)
+			log.Printf("#%d commented: %s", prNumber, testComment.GetHTMLURL())
 		}
-		log.Printf("#%d was retested %d times", prNumber, retested)
-		if retested > 3 {
+
+		if len(jobsToRetest) != 0 {
 			continue
 		}
 
-		prDetails, _, err := client.PullRequests.Get(ctx, "stackrox", "stackrox", prNumber)
+		retestComment := github.IssueComment{
+			Body: ptr(retestComment),
+		}
+		if !shouldRetest(statuses, commentsBodies) {
+			continue
+		}
+		comment, _, err := client.Issues.CreateComment(ctx, S, S, prNumber, &retestComment)
 		handleError(err)
+		log.Printf("#%d commented: %s", prNumber, comment.GetHTMLURL())
+	}
+}
 
-		if isGHACheckFailing(ctx, client, prDetails.GetHead().GetSHA()) {
+var (
+	restestNTimes = regexp.MustCompile("Retest (.*) (\\d+) times")
+	testJob       = regexp.MustCompile("/test (.*)")
+)
+
+func jobsToRetestFromComments(comments []string) []string {
+	testedJobs := map[string]int{}
+	jobsToRetest := map[string]int{}
+
+	for _, c := range comments {
+		testJobMatch := testJob.FindStringSubmatch(c)
+		if len(testJobMatch) == 2 {
+			job := testJobMatch[1]
+			if _, ok := testedJobs[job]; !ok {
+				testedJobs[job] = 0
+			}
+			testedJobs[job]++
 			continue
 		}
 
-		var statuses []Status
-		statusRequest, err := http.NewRequest("GET", prDetails.GetStatusesURL(), nil)
-		handleError(err)
-		_, err = client.Do(ctx, statusRequest, &statuses)
-		handleError(err)
+		matched := restestNTimes.FindStringSubmatch(c)
+		if len(matched) != 3 {
+			continue
+		}
+		job := matched[1]
+		t, err := strconv.Atoi(matched[2])
+		if err != nil {
+			log.Printf("got an error in a comment %s: %s", c, err)
+			continue
+		}
+		if t < 1 || t > 100 {
+			log.Printf("invalid retest number requested: %s", c)
+			continue
+		}
+		if _, ok := jobsToRetest[job]; !ok {
+			jobsToRetest[job] = 0
+		}
+		jobsToRetest[job] += t
+	}
 
-		retestComment := github.IssueComment{
-			Body: &retestComment,
+	missingTests := make([]string, 0, len(jobsToRetest))
+	for job, times := range jobsToRetest {
+		toTest := times - testedJobs[job]
+		if toTest < 1 {
+			continue
 		}
+		missingTests = append(missingTests, job)
+	}
+	slices.Sort(missingTests)
 
-		for _, status := range statuses {
-			log.Printf("#%d %-40s\t%10s", prNumber, status.Context, status.State)
-			if status.State != "failure" {
-				continue
-			}
-			comment, _, err := client.Issues.CreateComment(ctx, "stackrox", "stackrox", prNumber, &retestComment)
-			handleError(err)
-			log.Printf("#%d commented: %s", prNumber, comment.GetURL())
-			break
+	return missingTests
+}
+
+const retestComment = "/retest"
+
+func shouldRetest(statuses map[string]string, comments []string) bool {
+
+	retested := 0
+	for _, c := range comments {
+		if c == retestComment {
```

Comment:
While common use, I think this command can appear in addition to other text, so `==` is not sufficient?

Reply:
Maybe it can but comments issued by this tool are exactly like that. This logic was not changed

Reply:
Yes, but a non-bot user can also issue these commands. It would be surprising if the bot counts such commands from any user, but only if they match a certain form.

Reply:
Right, fixed